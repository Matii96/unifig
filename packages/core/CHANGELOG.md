# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2022-10-12)

### Bug Fixes

- **core:** added missing class-validator dependency ([6052855](https://github.com/Matii96/unifig/commit/60528552db7396b967907cd262cecb3612f7293a))

### Features

- **core:** improved manager interface ([cef08c7](https://github.com/Matii96/unifig/commit/cef08c77b380c5c063969821829adc309705f927))
- **core:** package setup ([cb2a066](https://github.com/Matii96/unifig/commit/cb2a066f4d40dfa31779cb187067f7134204ac3b))
- **core:** throwing exceptions upon encountering problems ([86b48d7](https://github.com/Matii96/unifig/commit/86b48d7166ba79a75755ade0a791271277d7c521))
- **env:** package setup ([5221aef](https://github.com/Matii96/unifig/commit/5221aeff7ac7b071619be5f6c9e537bb74cdacb8))

### BREAKING CHANGES

- **core:** changed getters names to more descriptive equivalents
