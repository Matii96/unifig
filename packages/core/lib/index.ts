export * from './adapters';
export * from './manager';
export * from './properties-mapping';
export * from './validator';
