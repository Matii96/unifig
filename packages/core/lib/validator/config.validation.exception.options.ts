import { ValidationError } from 'class-validator';
import { Type } from '../utils/type.interface';

export interface ConfigValidationExceptionOptions {
  template: Type<any>;
  errors: ValidationError[];
}
