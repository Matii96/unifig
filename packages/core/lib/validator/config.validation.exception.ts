export class ConfigValidationException<TConfig extends object> extends Error {
  constructor(config: TConfig, msg: string) {
    super(`Validation failed for ${config.constructor.name}:${msg}`);
    this.name = ConfigValidationException.name;
  }
}
