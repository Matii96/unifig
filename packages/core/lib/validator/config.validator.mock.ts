export class ConfigValidatorMock {
  validate = jest.fn();
}
