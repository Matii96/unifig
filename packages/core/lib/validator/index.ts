export * from './decorators';
export * from './errors';
