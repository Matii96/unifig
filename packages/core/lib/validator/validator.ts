import { ConfigValidationError } from './errors/validation.error';

export interface Validator {
  validate(configs: object[]): ConfigValidationError | undefined;
}
