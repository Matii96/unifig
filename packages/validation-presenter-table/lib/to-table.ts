import {
  ConfigTemplateValidationError,
  ConfigValidationError,
  ConfigPropertyValidationError,
  ConfigSubtemplateValidationError,
} from '@unifig/core';
import { getBorderCharacters, SpanningCellConfig, table } from 'table';
import { TemplateRow } from './template-row/template-row';
import { ToTableOptions } from './to-table.options';

const HEADER = ['Template', 'Property', 'Type', 'Source', 'Current Value', 'Failed constraints'];

export const toTable = (
  validationException: ConfigValidationError,
  options: ToTableOptions = {}
): string => {
  const tableRows = formatTableRowsGroup(validationException.errors);
  const tableData = [HEADER, ...tableRows.flatMap(({ templateTableData }) => templateTableData)];

  return table(tableData, {
    columns: [{ alignment: 'left', width: 16 }],
    spanningCells: tableRows.map(({ spanningCells }) => spanningCells),
    border: getBorderCharacters(options.border ?? 'norc'),
  });
};

const formatTableRowsGroup = (failedValidations: ConfigTemplateValidationError[]) => {
  let spanningCellsRowIdx = 1;
  return failedValidations.map((failedValidation) => {
    const templateTableData = failedValidation.errors
      .flatMap((error) => formatTemplateRows(error))
      .map((row) => row.toArray());
    templateTableData[0][0] = failedValidation.template.name;

    const spanningCells: SpanningCellConfig = {
      col: 0,
      row: spanningCellsRowIdx,
      rowSpan: templateTableData.length,
      verticalAlignment: 'middle',
    };
    spanningCellsRowIdx += templateTableData.length;
    return { templateTableData, spanningCells };
  });
};

const formatTemplateRows = (
  error: ConfigPropertyValidationError | ConfigSubtemplateValidationError,
  parentPrefix = ''
): TemplateRow[] => {
  if (error instanceof ConfigPropertyValidationError) {
    return [TemplateRow.fromValidationError(error, parentPrefix)];
  }
  if (error instanceof ConfigSubtemplateValidationError) {
    const propertyParentPrefix = parentPrefix + error.property + '.';
    const subtemplateRows = error.children.flatMap((child) =>
      formatTemplateRows(child, propertyParentPrefix)
    );
    return error.failedConstraints
      ? [TemplateRow.fromValidationError(error), ...subtemplateRows]
      : subtemplateRows;
  }
  throw new Error('Passed plain validation object: ' + JSON.stringify(error));
};
