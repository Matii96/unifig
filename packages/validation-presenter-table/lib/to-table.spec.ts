import { ConfigValidationError } from '@unifig/core';
import { mockFailed2dValidation } from './validator.mocks';
import { toTable } from './to-table';

describe('toTable', () => {
  it('should format table from 2d errors', () => {
    expect(toTable(new ConfigValidationError(mockFailed2dValidation()))).toMatchSnapshot();
  });
});
