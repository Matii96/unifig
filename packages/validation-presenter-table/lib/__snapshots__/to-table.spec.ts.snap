// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toTable should format table from 1d error 1`] = `
"┌──────────────────┬──────────┬────────┬────────┬───────────────┬────────────────────┐
│ Template         │ Property │ Type   │ Source │ Current Value │ Failed constraints │
├──────────────────┼──────────┼────────┼────────┼───────────────┼────────────────────┤
│ StorageOptions   │ port     │ Number │ PORT   │ "not-a-port"  │ isInt              │
└──────────────────┴──────────┴────────┴────────┴───────────────┴────────────────────┘
"
`;

exports[`toTable should format table from 2d errors 1`] = `
"┌──────────────────┬─────────────┬────────┬─────────────┬───────────────┬────────────────────┐
│ Template         │ Property    │ Type   │ Source      │ Current Value │ Failed constraints │
├──────────────────┼─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│                  │ port        │ Number │ PORT        │ "not-a-port"  │ isInt              │
│                  ├─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│ StorageOptions   │ db.url      │ String │ DB_URL      │ undefined     │ isString           │
│                  ├─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│                  │ db.password │ String │ DB_PASSWORD │ undefined     │ isString           │
├──────────────────┼─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│ NetworkOptions   │ ipRange     │        │             │ undefined     │ isDefined          │
└──────────────────┴─────────────┴────────┴─────────────┴───────────────┴────────────────────┘
"
`;
